dataplane/standalone/saiserver/ports.go

```diff
@@ -0,0 +1,195 @@
+	ID() string
+	CreatePort(ctx context.Context, req *dpb.CreatePortRequest) (*dpb.CreatePortResponse, error)
+	PortState(ctx context.Context, req *fwdpb.PortStateRequest) (*fwdpb.PortStateReply, error)
+}
+
+func newPort(mgr *attrmgr.AttrMgr, dataplane portDataplaneAPI, s *grpc.Server) *port {
+	p := &port{
+		mgr:       mgr,
+		dataplane: dataplane,
+		portToEth: make(map[uint64]string),
+	}
+	saipb.RegisterPortServer(s, p)
+	return p
+}
+
+type port struct {
+	saipb.UnimplementedPortServer
+	mgr       *attrmgr.AttrMgr
+	dataplane portDataplaneAPI
+	nextEth   int
+	portToEth map[uint64]string
+}
+
+// stub for testing
+var getInterface = net.InterfaceByName
+
+// CreatePort creates a new port, mapping the port to ethX, where X is assigned sequentially from 1 to n.
+// Note: If more ports are created than eth devices, no error is returned, but the OperStatus is set to NOT_PRESENT.
+func (port *port) CreatePort(ctx context.Context, _ *saipb.CreatePortRequest) (*saipb.CreatePortResponse, error) {
+	id := port.mgr.NextID()
+	port.nextEth++
+	dev := fmt.Sprintf("eth%v", port.nextEth)
+
+	attrs := &saipb.PortAttribute{
+		OperSpeed:                     proto.Uint32(1024),
+		NumberOfIngressPriorityGroups: proto.Uint32(0),
+		QosNumberOfQueues:             proto.Uint32(0),
+		QosMaximumHeadroomSize:        proto.Uint32(0),
+		AdminState:                    proto.Bool(true),
+		AutoNegMode:                   proto.Bool(false),
+		Mtu:                           proto.Uint32(1514),
+	}
+
+	// For ports that don't exist, do not create dataplane ports.
+	if _, err := getInterface(dev); err != nil {
+		attrs.OperStatus = saipb.PortOperStatus_PORT_OPER_STATUS_NOT_PRESENT.Enum()
+		port.mgr.StoreAttributes(id, attrs)
+		return &saipb.CreatePortResponse{
+			Oid: id,
+		}, nil
+	}
+	port.portToEth[id] = dev
+
+	_, err := port.dataplane.CreatePort(ctx, &dpb.CreatePortRequest{
+		Id:   fmt.Sprint(id),
+		Type: fwdpb.PortType_PORT_TYPE_KERNEL,
+		Src: &dpb.CreatePortRequest_KernelDev{
+			KernelDev: dev,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	attrs.OperStatus = saipb.PortOperStatus_PORT_OPER_STATUS_UP.Enum().Enum()
+	port.mgr.StoreAttributes(id, attrs)
+
+	return &saipb.CreatePortResponse{
+		Oid: id,
+	}, nil
+}
+
+// SetPortAttributes sets the attributes in the request.
+func (port *port) SetPortAttribute(ctx context.Context, req *saipb.SetPortAttributeRequest) (*saipb.SetPortAttributeResponse, error) {
+	if req.AdminState != nil {
+		stateReq := &fwdpb.PortStateRequest{
+			ContextId: &fwdpb.ContextId{Id: port.dataplane.ID()},
+			PortId:    &fwdpb.PortId{ObjectId: &fwdpb.ObjectId{Id: fmt.Sprint(req.GetOid())}},
+		}
+		stateReq.Operation = &fwdpb.PortInfo{
+			AdminStatus: fwdpb.PortState_PORT_STATE_DISABLED_DOWN,
+		}
+		if req.GetAdminState() {
+			stateReq.Operation.AdminStatus = fwdpb.PortState_PORT_STATE_ENABLED_UP
+		}
+		_, err := port.dataplane.PortState(ctx, stateReq)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return nil, nil
```

Comment:
should this return

saipb.SetPortAttributeResponse{}, nil

? I know its basically the same thing

Reply:
done